### PR TITLE
fix(mcp): add descriptions to Notion MCP tool registrations

### PR DESCRIPTION
**Pull request by baboosh**
Notion MCP tools were registered without description strings, causing the LLM to receive tools with empty descriptions. Unlike GitHub and Atlassian, Notion also had no fallback descriptions in MCP_PROVIDER_METADATA. The LLM could not determine when or how to use the tools, so it never invoked them despite the integration being connected.

- Add description strings to all 4 Notion tool registrations (search, create_page, read_page, append_blocks)
- Add notion entry to MCP_PROVIDER_METADATA with fallback descriptions
- Update test utils to capture tool descriptions
- Add tests verifying all Notion tools have non-empty descriptions

# Pull Request

## Security Testing (for security-labeled PRs only)

- [ ] Vulnerability reproduced on **staging** with a script/curl (output attached as PR comment)
- [ ] Fix verified on **PR preview** env with the same script (output attached as PR comment)
- [ ] Production is assumed to match staging (no direct-to-prod deploys). If BEFORE reproduction on staging fails unexpectedly, verify no upstream fix has already landed: `git log origin/main -- <file>`

## Checklist

- [ ] I have added the new AdminSettings to Staging, prod and the hydration script (if applicable)
- [ ] I have made the UI/UX feel good (toasters, size, responsive, etc)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (if applicable)
- [ ] My changes generate no new warnings
- [ ] If adding/modifying telemetry fields: updated [telemetry data classification](../docs-site/docs/security/telemetry-data-classification.md) doc

### CODE REVIEW

**Comment by wescarda** [comment]
## MCP Playwright Test - PR #1979 (add descriptions to Notion MCP tool registrations)

**URL:** N/A - no preview deployed (`app.pr1979.preview.bike4mind.com` is NXDOMAIN)
**Date:** 2026-08-20
**Method:** Code + test verification (backend-only MCP metadata PR; no UI surface, no Guide for Testers, no preview - nothing for Playwright to exercise)

---

## Verdict

> ✅ **PASS (code-level) - no visual surface to Playwright-test.**
> This PR only changes MCP tool-registration metadata (description strings the LLM receives); there's no rendered UI, the Guide for Testers is the empty template, and no preview exists. Verified by diff + the PR's own tests + CI, all of which check out.

---

## Summary table

| # | Test step | Result |
|---|---|---|
| 1 | Visual/Playwright surface | N/A - backend-only (MCP tool metadata) |
| 2 | Guide for Testers present | N/A - empty PR template |
| 3 | Preview reachable | N/A - NXDOMAIN, not deployed |
| 4 | All 4 Notion tools registered with a description | ✅ Pass - `search.ts` +1, `pages.ts` +3 (create/append/read) |
| 5 | `MCP_PROVIDER_METADATA` gains a `notion` fallback | ✅ Pass - `providers.ts` adds `notion.defaultToolDescriptions` (4 tools) |
| 6 | Tests assert non-empty descriptions | ✅ Pass - new `search.test.ts` + `pages.test.ts`; `test-utils.ts` captures the description arg |
| 7 | CI | ✅ 11/18 complete, 0 failures (build/lint/test shards still running) |

---

## What works

- All four Notion tool registrations now pass a concrete description (`notion_search`, `notion_create_page`, `notion_append_blocks`, `notion_read_page`).
- `providers.ts` adds a `notion` entry to `MCP_PROVIDER_METADATA` with `defaultToolDescriptions` for all four - closing the gap vs GitHub/Atlassian, which already had fallbacks.
- `test-utils.ts` updated to capture the registration description; new tests assert every Notion tool has a truthy string description.
- Changeset included; CI green so far with no failures.

---

## Other observations (not blockers)

- Not a visual/Playwright PR - no screenshots taken, output folder not opened.
- The real-world payoff (the LLM now actually invokes the Notion tools) needs a live agent run against a connected Notion workspace; that's out of scope for a static/visual pass and not reproducible without a preview + Notion credentials.
- The PR's structured template sections (Description/Changes/Guide for Testers) are left as empty comments; a filled Guide for Testers would help a human confirm the agent-invocation payoff.

---

No screenshots - backend-only PR with no UI surface or preview.

File: b4m-core/common/src/mcp/providers.ts [hunk 1]
```diff
@@ -69,6 +69,18 @@ const atlassianDescriptions: Record<string, string> = {
 };
 
 export const MCP_PROVIDER_METADATA: Record<string, McpProviderMetadata> = {
+  notion: {
```
Comment (vinchi777) [inline b4m-core/common/src/mcp/providers.ts:74]: These four strings are byte-for-byte copies of the descriptions now passed to `server.tool(...)` in `notion/tools/search.ts` and `notion/tools/pages.ts`. Nothing keeps them in sync: the fallback is what the admin tool-definitions endpoints render and what `generateMcpToolsFromCache` hands the LLM for a cached tool list, while the registration string is what a live `getTools()` returns. Edit one and the two silently diverge, so the same tool ends up described differently depending on which path served it.

The `github` block below at least carries a `// Core GitHub tools (must match b4m-core/mcp/src/github/index.ts)` comment. Either add the equivalent pointer here, or better, export the description strings from `b4m-core/mcp/src/notion/constants.ts` next to the `TOOL_NOTION_*` names and reference them from both sides so there is one source of truth.

File: b4m-core/mcp/src/notion/__tests__/tools/pages.test.ts [hunk 2]
```diff
@@ -70,6 +70,14 @@ describe('Page Tools', () => {
     registerPageTools(mock.server);
   });
 
+  it('all page tools should have non-empty descriptions', () => {
```
Comment (vinchi777) [inline b4m-core/mcp/src/notion/__tests__/tools/pages.test.ts]: `toBeTruthy()` plus a `typeof` check catches the original regression (empty description) but not the one this PR newly introduces: drift between the registration string and the `MCP_PROVIDER_METADATA.notion.defaultToolDescriptions` copy. Asserting `tool.description` equals the fallback entry for the same tool name would lock the invariant that actually matters and costs one line. The `typeof` assertion is also redundant - `RegisteredTool.description` is `string | undefined`, so `toBeTruthy()` already rules out everything but a non-empty string.

File: b4m-core/mcp/src/notion/tools/pages.ts [hunk 3]
```diff
@@ -563,6 +565,7 @@ export function registerPageTools(server: McpServer): void {
   // --- Read Page ---
   server.tool(
     TOOL_NOTION_READ_PAGE,
+    'Read the content of a Notion page by its ID. Returns the child blocks (text, headings, lists, etc.) and a plain-text summary.',
```
Comment (vinchi777) [inline b4m-core/mcp/src/notion/tools/pages.ts]: The handler defaults `page_size` to 100 and returns `has_more`/`next_cursor`, but the description does not say the result is paginated. A model that reads a truncated page gets no cue from the tool description that more blocks exist - worth a clause like "Results are paginated; pass start_cursor with the returned next_cursor when has_more is true." Take it or leave it.

**Review by vinchi777 — DISMISSED** [review]
| Severity | Finding | Location |
|----------|---------|----------|
| P2 | Notion descriptions are duplicated verbatim between the tool registrations and the `MCP_PROVIDER_METADATA` fallback, with no cross-reference comment and no test pinning them together | `b4m-core/common/src/mcp/providers.ts:72` |
| P3 | New tests assert only that a description is non-empty, not that it matches the declared fallback | `b4m-core/mcp/src/notion/__tests__/tools/pages.test.ts:73` |
| P3 | `notion_read_page` description omits that the result is paginated | `b4m-core/mcp/src/notion/tools/pages.ts:568` |

Diagnosis and fix both check out. The four `defaultToolDescriptions` keys match the `TOOL_NOTION_*` constant values exactly (`notion_search`, `notion_create_page`, `notion_read_page`, `notion_append_blocks`), and the metadata key `notion` matches the `McpServer` name in `b4m-core/mcp/src/notion/index.ts`, so `getMcpProviderMetadata` resolves for both `generateMcpTools` and `generateMcpToolsFromCache` - the latter being what actually rescues sessions holding a cached tool list with empty descriptions. Descriptions are accurate against the handlers: create_page and append_blocks do gate on `checkWriteAccess()`, read_page does return `blocks` plus `plain_text`, and search does return `title`/`id`/`url` for pages and databases. The `createMockServer` change reads the description off `args[0]` only when it is a string, which is correct for both `tool(name, schema, cb)` and `tool(name, description, schema, cb)`. Adding a `notion` entry also surfaces these four tools in the admin tool-definitions endpoints; the `${providerName}_${toolId}` composition yields `notion_notion_search`, which is ugly but round-trips correctly through the prefix strip in `[toolId].ts` and matches the existing `github_github_*` precedent.

CI: the only failing job is `server/services/deleteOrganizationTransaction.e2e.test.ts` timing out at 60s in the client shard - unrelated to this diff and flaky against the replica set. The `b4m-core/mcp` suite (441 tests, including the new ones) passes in the misc shard.

**Review by ken-b4m — DISMISSED** [review]
## Review: fix(mcp) — add descriptions to Notion MCP tool registrations

**CI:** Green at head `1448a0fd`, resolved per check name against the head SHA — only `E2E Tests`, `Merge multi-arch manifest` and `review` skipped, nothing failing.
**Scope:** whole PR, **+69/-6** across 8 files. `reviewDecision: REVIEW_REQUIRED`, no prior reviews, no threads — first review.

---

### Overall Assessment: **APPROVE** — the descriptions are accurate against the implementations, which is the thing that actually matters here. One P3: they now exist in two places with nothing keeping them equal.

**A missing tool description is a real defect, not a polish item.** These strings are what a model reads to decide whether to call a tool and with what arguments, so an unregistered description means the model is choosing between `notion_search` and `notion_read_page` on their names alone. Fixing it is worth more than its diff size suggests.

**The descriptions match what the code does — I checked the one that makes a specific promise.** `notion_read_page` tells the model "Results are paginated; pass `start_cursor` with the returned `next_cursor`", which is exactly the class of claim that misleads a model if the parameter does not exist. It does: `pages.ts:583` declares `start_cursor: startCursorSchema`, `:585` destructures it, and `:594-595` forwards it to the Notion API. Same for the write-access preconditions on `notion_create_page` and `notion_append_blocks` — stating a precondition in the description is the right place for it, since the model can then avoid a call it cannot complete rather than learning from an error.

**The test-harness change is the right shape for the change it supports.** `createMockServer` now records the description by sniffing the argument list (`typeof args[0] === 'string'`), with a comment naming both real call shapes — `tool(name, description, schema, handler)` and `tool(name, schema, handler)`. That keeps the mock able to assert descriptions without forcing every existing registration to grow one, which is what makes the incremental adoption possible.

---

### P3 — Minor

**The Notion descriptions now exist twice, feeding two different consumers, kept in step only by a comment.** `b4m-core/common/src/mcp/providers.ts` and `b4m-core/mcp/src/notion/constants.ts`

The comment is honest about it — "Must stay in sync with `TOOL_DESCRIPTIONS` ... (common cannot import from mcp due to the dependency direction)" — and the dependency argument is correct. I verified the two copies are byte-identical today, so nothing is wrong at head.

What makes this worth raising is that the copies are not two views of one thing; they are read by different layers:

- `TOOL_DESCRIPTIONS` drives the live `server.tool(name, description, …)` registrations (`tools/pages.ts:411,515,573`, `tools/search.ts:137`) — what the MCP server advertises to a connected client.
- `MCP_PROVIDER_METADATA.notion.defaultToolDescriptions` is read by the admin tool-definitions endpoints **and** by `services/src/llm/tools/toolGenerators.ts:192,275` as `fallbackDescription` — what the LLM is told when no override exists.

So a drift does not produce a cosmetic mismatch: the model would be told one thing while the MCP server advertises another, and the admin UI would show the model's version. On `notion_create_page`, where the description carries the write-access precondition, that is the kind of divergence someone debugs for a while.

There is no test holding them together. The one file that references `MCP_PROVIDER_METADATA` in a test — `apps/client/__tests__/tool-definitions.test.ts:43` — mocks it as `{}`, so it asserts nothing about the contents.

The clean home for the guard is `b4m-core/mcp`, which *can* import `@bike4mind/common` (that is the direction the comment describes), so a test there can hold both and assert equality — including that the key sets match, so a fifth tool added to one side fails the build rather than silently arriving without a description on the other. Two or three lines, and it converts the comment into something enforced.

---

<sub>Reviewed at `1448a0fd`. CI resolved per check name against the head SHA. The descriptions were checked against the implementations rather than read for plausibility — specifically `start_cursor`'s existence on the read-page schema, since a description promising a parameter the tool lacks actively misleads the model. The two copies were compared verbatim to establish that this is a drift risk rather than a live mismatch, and each one's consumers were traced to confirm they feed different layers. The absence of a sync test was verified by reading the one test that references the metadata. **Stated limitation: no tests were run and no MCP client was connected** — the registration behaviour is from reading the `server.tool` call sites.</sub>

**Review by ken-b4m — APPROVED** [review]
## Re-review: fix(mcp) — add descriptions to Notion MCP tool registrations

**Reviewed at:** `49499ecb` (my APPROVE was at `1448a0fd`, dismissed by the push). Delta is one commit and one file: `description-sync.test.ts`, +17 lines.
**CI:** Green at this head — only the usual `E2E Tests`, `Merge multi-arch manifest` and `review` skips.
**Prior reviews:** mine, plus three threads from @vinchi777 — all adjudicated below. **One is still open and blocks the merge** under `main`'s thread-resolution rule.

---

### Verdict: **APPROVE** — my P3 is closed, and closed with both halves of what it asked for.

**The sync test does the thing that makes it worth having.** I asked for two properties and got both: `:9` asserts the **key sets** match (`Object.keys(fallback).sort()` against `Object.keys(TOOL_DESCRIPTIONS).sort()`), so a fifth Notion tool added to one side fails the build rather than silently shipping without a description; and `:12-16` asserts per-tool string equality, parameterized over `ALL_TOOL_NAMES` so the case list grows by itself.

Placing it in `b4m-core/mcp` is right — that package can import `@bike4mind/common` while the reverse would invert the dependency graph, which is the constraint that forced the duplication in the first place. The comment in `providers.ts` now has a test behind it instead of being the only thing holding the two copies together.

One structural detail worth noting because it is easy to get wrong: `const fallback = getMcpProviderMetadata('notion')?.defaultToolDescriptions ?? {}` could have made this vacuous — an accessor returning `undefined` would leave `fallback` empty and a per-tool loop over an empty object would assert nothing. It does not, because `:9` compares key sets, so an absent or renamed provider entry fails there. The assertion order is what saves it.

### @vinchi777's threads

- **`providers.ts` — "these four strings are byte-for-byte copies and nothing keeps them in sync" — FULLY CLOSED, and it needs resolving to unblock the merge.** This is the same finding I raised, arrived at independently, and `49499ecb` closes it. Worth noting they named a consumer I did not: `generateMcpToolsFromCache` hands the fallback to the LLM, alongside the admin tool-definitions endpoints — which is exactly why the two copies feeding *different* layers made the drift consequential rather than cosmetic. This thread is the only one still unresolved, and `main` enforces thread resolution, so it is what is holding the PR.
- **`pages.test.ts` — assert the registration string equals the fallback entry — SATISFIED TRANSITIVELY, worth one sentence.** The registrations pass the constant itself (`pages.ts:411,515,573`, `search.ts:137` all pass `TOOL_DESCRIPTIONS[...]`), so registration-to-constant drift is structurally impossible, and the new test ties constant to fallback. The chain therefore holds end to end. I flag it only because that argument depends on the registrations continuing to pass the constant rather than an inlined string — a one-line change away from being false, and the test in `pages.test.ts` that captures the mock's `description` is where you would notice. Not worth adding now.
- **`pages.ts` — the read-page description does not say results are paginated — FULLY CLOSED.** It now reads "Results are paginated; pass `start_cursor` with the returned `next_cursor`…", and I verified the parameter it promises is real: `start_cursor` is on the schema, destructured, and forwarded to the API. A description promising a parameter the tool lacks is the version of this that actively misleads a model, so this one mattered more than its size.

No new defects in the delta — it adds a test file and touches nothing else.

---

<sub>Re-reviewed at `49499ecb` against my prior `1448a0fd`; the delta was confirmed to be a single new test file before assessing anything else. The sync test was checked for vacuity (whether the `?? {}` fallback could let it pass with an empty object) rather than read for intent. The transitive-closure argument on @vinchi777's second thread was checked against the actual registration call sites, not assumed. **Stated limitation: no tests were run** — the new file was assessed by reading its assertions.</sub>
